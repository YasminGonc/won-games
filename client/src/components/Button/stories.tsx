import { StoryObj, Meta } from '@storybook/react'
import { ShoppingCart } from 'phosphor-react'
import { Button, ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  args: {
    children: 'Buy now',
    size: 'medium',
    fullWidth: false,
    minimal: false
  },
  argTypes: {
    icon: {
      type: 'symbol'
    }
  }
} as Meta<ButtonProps>

export const Default: StoryObj<ButtonProps> = {}

export const withIcon: StoryObj<ButtonProps> = {
  render: (args) => (
    <Button {...args}>
      <ShoppingCart />
      Buy now
    </Button>
  )
}

export const asLink: StoryObj<ButtonProps> = {
  args: {
    size: 'large',
    as: 'a',
    href: '/link',
    fullWidth: false
  }
}

export const minimal: StoryObj<ButtonProps> = {
  render: (args) => (
    <Button {...args}>
      <ShoppingCart />
      Buy now
    </Button>
  )
}

minimal.args = {
  minimal: true
}
