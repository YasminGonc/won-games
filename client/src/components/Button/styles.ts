import { darken } from 'polished'
import styled, { css, DefaultTheme } from 'styled-components'
import { ButtonProps } from '.'

const wrapperModifiers = {
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  minimal: (theme: DefaultTheme) => css`
    background: none;
    color: ${theme.colors.primary};

    &:hover {
      color: ${darken(0.1, theme.colors.primary)};
    }
  `
}

export const Wrapper = styled.button<ButtonProps>`
  ${({ theme, size, fullWidth, minimal }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    gap: 0.7rem;

    background: linear-gradient(
      178.59deg,
      #ff5f5f -14.51%,
      #f062c0 102.86%,
      #f23131 102.86%
    );
    border: none;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    cursor: pointer;

    color: ${theme.colors.white};
    text-decoration: none;

    ${!!size && wrapperModifiers[size](theme)};
    ${!!fullWidth && wrapperModifiers.fullWidth};
    ${!!minimal && wrapperModifiers.minimal(theme)}

    &:hover {
      background: ${minimal
        ? 'none'
        : `linear-gradient(180deg, #e35565 0%, #d958a6 100%)`};
    }
  `}
`
