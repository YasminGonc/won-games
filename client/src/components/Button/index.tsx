import {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  ElementType,
  ReactNode
} from 'react'
import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  icon?: ReactNode
  as?: ElementType
} & ButtonTypes

export function Button({
  children,
  size = 'medium',
  fullWidth = false,
  icon,
  ...props
}: ButtonProps) {
  return (
    <S.Wrapper size={size} fullWidth={fullWidth} {...props}>
      {icon}
      {children}
    </S.Wrapper>
  )
}
