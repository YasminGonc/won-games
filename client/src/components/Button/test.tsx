import { renderWithTheme } from '../../utils/tests/helpers'
import { screen } from '@testing-library/react'
import { ShoppingCart } from 'phosphor-react'

import Button from '.'

describe('<Button />', () => {
  it('should render a medium button by default', () => {
    const { container } = renderWithTheme(<Button>Buy now</Button>)
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a small button', () => {
    renderWithTheme(<Button size="small">Buy now</Button>)
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })

  it('should render a large button', () => {
    renderWithTheme(<Button size="large">Buy now</Button>)
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.6rem'
    })
  })

  it('should render a full width button', () => {
    renderWithTheme(<Button fullWidth>Buy now</Button>)
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<ShoppingCart data-testid="icon" />}>Buy now</Button>
    )
    expect(screen.getByText(/buy now/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render a minimal version', () => {
    renderWithTheme(
      <Button icon={<ShoppingCart data-testid="icon" />} minimal>
        Buy now
      </Button>
    )

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      background: 'none',
      color: '#F231A5'
    })

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyleRule(
      'background',
      'none',
      {
        modifier: ':hover'
      }
    )
  })

  it('should render button as a link', () => {
    const { debug, container } = renderWithTheme(
      <Button as="a" href="/link">
        Buy now
      </Button>
    )
    debug(container) //imprime o componente renderizado
    expect(screen.getByRole('link', { name: /buy now/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
