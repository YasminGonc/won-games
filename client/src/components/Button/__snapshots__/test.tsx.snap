// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a medium button by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 0.7rem;
  background: linear-gradient( 178.59deg, #ff5f5f -14.51%, #f062c0 102.86%, #f23131 102.86% );
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  cursor: pointer;
  color: #FAFAFA;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 100%);
}

<button
  class="c0"
>
  Buy now
</button>
`;
